### PR TITLE
[PTCD-636] Add Azure Function for fixing corrupt course status data

### PR DESCRIPTION
**Pull request by timabell**
Add Azure Function for fixing corrupt course status data

The source of the corruption has already been fixed in https://github.com/SkillsFundingAgency/dfc-providerportal-courses/pull/171

Test locally with

```bash
curl -v http://localhost:7071/admin/functions/FixCorruptStatusData -d "{}" -H "Content-Type:application/json"
```

[PTCD-636](https://skillsfundingagency.atlassian.net/browse/PTCD-636)

### CODE REVIEW

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        [FunctionName(nameof(FixCorruptStatusData))]
+        [NoAutomaticTrigger]
+        public async Task Run(string input)
+        {
+            var collectionUri = UriFactory.CreateDocumentCollectionUri(
+                _configuration.DatabaseId,
+                _configuration.CoursesCollectionName);
+
+            const CourseStatus courseRunStatus = CourseStatus.Live;
+            const CourseStatus unwantedCourseStatus = CourseStatus.BulkUploadReadyToGoLive;
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:41]: Unnecessary const

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            DocumentClient documentClient)
+        {
+            _configuration = configuration;
+            _documentClient = documentClient;
+            _logger = loggerFactory.CreateLogger<FixCorruptStatusData>();
+        }
+
+        [FunctionName(nameof(FixCorruptStatusData))]
+        [NoAutomaticTrigger]
+        public async Task Run(string input)
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:34]: `input` is unsed

File: src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs [hunk 3]
```diff
@@ -10,6 +10,11 @@ public class Course
     {
         [JsonProperty("id")]
         public Guid Id { get; set; }
+        /// <summary>
+        /// CosmosDb ETag for optimistic concurrency
+        /// https://github.com/Azure/azure-cosmos-dotnet-v2/blob/master/samples/code-samples/DocumentManagement/Program.cs
+        /// </summary>
+        public string _etag { get; set; }
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs:17]: Unnecessary comment?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                {
+                    course.CourseStatus = courseRunStatus;
+
+                    var documentUri = UriFactory.CreateDocumentUri(
+                        _configuration.DatabaseId,
+                        _configuration.CoursesCollectionName,
+                        course.Id.ToString());
+
+                    _logger.LogDebug($"Updating course with correct status {documentUri}");
+                    await _documentClient.ReplaceDocumentAsync(documentUri, course,
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:62]: Is there not an upsert document?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            Configuration configuration,
+            ILoggerFactory loggerFactory,
+            DocumentClient documentClient)
+        {
+            _configuration = configuration;
+            _documentClient = documentClient;
+            _logger = loggerFactory.CreateLogger<FixCorruptStatusData>();
+        }
+
+        [FunctionName(nameof(FixCorruptStatusData))]
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:32]: Ambiguous naming, consider something more descriptive? (even if it feels excessively verbose)

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 1 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:41]: Also, what do you mean by `unwanted`? Why don't we want it?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+                    var documentUri = UriFactory.CreateDocumentUri(
+                        _configuration.DatabaseId,
+                        _configuration.CoursesCollectionName,
+                        course.Id.ToString());
+
+                    _logger.LogDebug($"Updating course with correct status {documentUri}");
+                    await _documentClient.ReplaceDocumentAsync(documentUri, course,
+                        new RequestOptions
+                            {AccessCondition = new AccessCondition {Type = AccessConditionType.IfMatch, Condition = course._etag}});
+                    fixCount++;
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:65]: Does `ProcessAll` run with any degree of parallelism? Consider `Interlocked.Increment`.

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+            var query = _documentClient
+                .CreateDocumentQuery<Course>(collectionUri,
+                    new FeedOptions() {EnableCrossPartitionQuery = true})
+                .Where(c => c.CourseStatus == unwantedCourseStatus)
+                .AsDocumentQuery();
+
+            int fixCount = 0;
+            await query.ProcessAll(async courses =>
+            {
+                foreach (var course in courses.Where(c => c.CourseRuns.All(r => r.RecordStatus == courseRunStatus)))
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:52]: Can you not combine this with the initial query?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+        [FunctionName(nameof(FixCorruptStatusData))]
+        [NoAutomaticTrigger]
+        public async Task Run(string input)
+        {
+            var collectionUri = UriFactory.CreateDocumentCollectionUri(
+                _configuration.DatabaseId,
+                _configuration.CoursesCollectionName);
+
+            const CourseStatus courseRunStatus = CourseStatus.Live;
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:40]: This makes the intent of the code harder to read (IMO)

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 5 above)
Comment (soharadesign) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:32]: `SyncCourseWithCourseRunStatus`?

Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:32]: I think it needs to be narrower. The intention is _For all courses of status BulkUploadReadyToGoLive that have courseruns that are all Live then update the course status to live_. Who knows how to express that in a meaningful sentence mind you!

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+            const CourseStatus unwantedCourseStatus = CourseStatus.BulkUploadReadyToGoLive;
+
+            var query = _documentClient
+                .CreateDocumentQuery<Course>(collectionUri,
+                    new FeedOptions() {EnableCrossPartitionQuery = true})
+                .Where(c => c.CourseStatus == unwantedCourseStatus)
+                .AsDocumentQuery();
+
+            int fixCount = 0;
+            await query.ProcessAll(async courses =>
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:50]: Does this need to be done in a `ProcessAll`? If you enumerate `query` you can just iterate using a `foreach`?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 4 above)
Comment (soharadesign) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:62]: I've seen situations where UpsertDocumentAsync creates a new doc with the same ID if you are using partition keys. Never really understood why, so I tend to favour ReplaceDocumentAsync myself!

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+            var collectionUri = UriFactory.CreateDocumentCollectionUri(
+                _configuration.DatabaseId,
+                _configuration.CoursesCollectionName);
+
+            const CourseStatus courseRunStatus = CourseStatus.Live;
+            const CourseStatus unwantedCourseStatus = CourseStatus.BulkUploadReadyToGoLive;
+
+            var query = _documentClient
+                .CreateDocumentQuery<Course>(collectionUri,
+                    new FeedOptions() {EnableCrossPartitionQuery = true})
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:45]: Do you need `MaxItemCount = -1` here?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 2 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:34]: https://github.com/Azure/azure-functions-vs-build-sdk/issues/168

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+using Microsoft.Azure.Documents.Linq;
+using Microsoft.Azure.WebJobs;
+using Microsoft.Extensions.Logging;
+
+namespace Dfc.CourseDirectory.Functions
+{
+    /// <summary>
+    /// In prod CourseStatus has got out of sync with the CourseRuns
+    /// </summary>
+    public class FixCorruptStatusData
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:16]: Bear in mind that this is not really testable over an isolated set of records.

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 1 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs]: agreed in retrospect, was trying to add explanation but agree it doesn't help
unwanted/incorrect/corrupt, shrug

File: src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs (hunk 3 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs:17]: I think it deserves some kind of pointer to it being a cosmos thing.

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 4 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:59]: Replace matches the intention, we don't want it creating courses

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 5 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:32]: Open to suggestions. I think it's good enough. Hopefully it'll be short-lived.

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 6 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:62]: Doesn't appear to https://github.com/SkillsFundingAgency/dfc-coursedirectory/blob/038ae36bf15e0dc188f416a091130bbe8b4981e3/src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/DocumentQueryExtensions.cs#L25

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 7 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs]: I tried, cosmos linq didn't support what I tried (`All()`), runtime failure
```
[2020-11-04T17:19:24.988] Executed 'FixCorruptStatusData' (Failed, Id=ad67db82-766b-4395-963c-79931c8554b8, Duration=10662ms)
[2020-11-04T17:19:24.988] System.Private.CoreLib: Exception while executing function: FixCorruptStatusData. Microsoft.Azure.DocumentDB.Core: Method 'All' is not supported., Linux/19.3 documentdb-netcore-sdk/2.9.2.
```
Gave up. It's not a huge amount of data after the first filter.

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 8 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs]: agreed in hindsight. gone.

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 10 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:42]: I think [FeedOptions.MaxItemCount](https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.documents.client.feedoptions.maxitemcount?view=azure-dotnet) sets the page size https://stackoverflow.com/a/52352814/10245
Then further down `ProcessAll` loops through the pages.
I *think* this means it's fine as is?

File: src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs (hunk 11 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Functions/FixCorruptStatusData.cs:16]: Discussed in tech chat - we could break out reading, add parameters for easier manual testing but opted to move fast on this one.

File: src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs (hunk 3 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs:17]: Fair enough. Can you not used a convention correct property name and `JsonProperty`?

**Review by jameslawesynergy — APPROVED** [review]

File: src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs (hunk 3 above)
Comment (timabell) [reply src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs:17]: Ah, fab, didn't think of that. Thanks! I'll do that

Comment (timabell) [reply src/Dfc.CourseDirectory.Core/DataStore/CosmosDb/Models/Course.cs]: done

**Comment by timabell** [comment]
[Release-882 for testing](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=17201&_a=release-pipeline-progress)

**Pull request merged**